### ISSUE TITLE
64-Bit Version performs worse than 32-Bit on x64 Windows 10 XPS13 Laptop

### ISSUE THREAD

**Issue by cwsoft** [tonioni/WinUAE#210]
Performance of 64-bit Version is much worse compared to the 32-bit Version on my 64-bit Dell XPS13 Laptop Windows 10 in latest release. Tested on A500 Kick 1.3 and A1200 Kick 3.1 and various Setups like AmigaOS 3.2.1, AmiKit 11.5, Scalos WB etc. Boot up time and screen updates are much slower than with identical setup and 32-bit WinUAE. 

Tested without any controller or external mouse etc. No glue whats going on here.

**Comment by tonioni** [comment]
Did older versions have similar slower speed? There shouldn't be any 64-bit vs 32-bit differences except JIT.

**Comment by tonioni** [comment]
Could you confirm if 4.9.0 and 4.4 has same problem? Download zip package and temporarily overwrite winuae.exe, there is no need to install anything.

**Comment by cwsoft** [comment]
Hello Toni. Sorry for the late replay, missed the notifications. Yepp, can confirm that all tested WinUAE 64-versions (v4.4.0, 4.9.0, 4.9.1) are much slower when it comes to loading the Workbench files. Some measurements on booting up AmiKitXE 11.7.0 (WB 45.3, KS: 40.68, Amiga 1200) gives the following times until the WB is fully loaded:
- v4.9.1 (32-bit): 14,7 seconds
- v4.9.1 (64-bit): 37,1 seconds
- v4.9.0 (64-bit): 36,7 seconds
- v4.4.0 (64-bit): 35,6 seconds

So basically all setups on my Win64 / XPS13 Laptop took about 20s longer when booting up distros like AmiKit, Pimiga, or my custom Workbench setups.

**Comment by cwsoft** [comment]
Hello Toni, have done some more tests. Previous tests were all with JIT enabled and booting into Full-Window.

**WinUAE 4.9.1 (both 32/64 bit variants):**
- 32-bit (no JIT, Full-Window): 28s
- 64-bit (no-JIT, Full-Window): 25s
- 32-bit (JIT, Fullscreen): 15s (don't see AmiKit Logo at startup, most likely boots too fast)
- 64-bit (JIT, Fullscreen): 38s (seeing AmiKit Logo at startup)

So disabling JIT and booting into Full-Window basically shows the same boot-up time for 32/64 bit.

**Comment by tonioni** [comment]
Thanks. It probably means 64-bit has always been "slow" in this situation. I am not sure if I can fix it because I am not that familiar with JIT (Also 64-bit variant was originally implemented by Aranym developers).

What if you change JIT cache size? (try both smaller and larger if it isn't already set to largest) 64-bit generated code is usually larger than 32-bit which could overflow JIT cache more easily and/or it might not anymore fit in host CPU caches which would also cause noticeable slowdown if 32-bit generated code still fits.

**Comment by cwsoft** [comment]
JIT cache size was set to max. for all my tests already. Fun fact somehow is that Amiberry v5.2-64-bit (which uses quite some WinUAE code) on my Pi400 with JIT enabled boots-up about as fast as the WinUAE 32-bit version on my Win10 Dell XPS Laptop. 

Do you know if Midwan (author of Amiberry) implemented JIT differently?

**Comment by midwan** [comment]
Amiberry uses a different implementation of JIT, since it's ARM-based. The JIT code comes from TomB originally (https://github.com/PandTomB/uae4arm), though I might make some minor changes here and there based on the updates coming from WinUAE.

**Comment by tonioni** [comment]
Did you try smaller JIT cache? It can be faster in some situations (as I said in previous post)
I assume CPU benchmarks returns similar results? (at least I get faster results in AIBB or sysinfo). To confirm that JIT does work as expected at least in situations where there is nothing else to do (like HD accesses).

**Comment by cwsoft** [comment]
Yepp. Tried max, min and in between. Numbers change marginally but 64-bit is always worse than 32-bit by 15-20s measured in boot-up time from a cold start to Workbench initialized and ready to use.

**Comment by tonioni** [comment]
This is quite interesting. What is your laptop CPU model? (full model information). I suspect it is CPU cache related.

**Comment by cwsoft** [comment]
Hi Toni. My Laptop CPU model is **Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz   2.40 GHz, 2 Cores, 4 logigal CPUs**.
Further details can be found in the attached dump-file of CPU-Z.
[dellxps13.txt](https://github.com/tonioni/WinUAE/files/8790529/dellxps13.txt)
